front/components/app/blocks/Input.tsx

```diff
@@ -40,12 +45,67 @@ export default function Input({
   onBlockDown: () => void;
   onBlockNew: (blockType: BlockType | "map_reduce" | "while_end") => void;
 }>) {
-  const handleSetDataset = (dataset: string) => {
+  const [datasetWithData, setDatasetWithData] = useState<DatasetType | null>(
+    null
+  );
+
+  const handleSetDataset = async (dataset: string) => {
     const b = shallowBlockClone(block);
     b.config.dataset = dataset;
+    setDatasetWithData(await handleGetDatasetData(dataset));
```

Comment:
This is a very slow call that we don't in the logic of adding/editing the Input block no?

Shuldn't this be an SWR hook that depends on b.config.dataset  ?